In [1]:
# Importing the Required Libraries
import pandas as pd
import numpy as np

In [2]:
# Reading TSV file 
df = pd.read_csv("T2 - uncleaned_data.tsv", sep="\t")

In [3]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

### 1. Missing Values:
#### Question: Check for missing values in each column (Order ID, Quantity, Item Name, Choice Description, Item Price). How should missing values be handled?

In [4]:
df.isnull().sum()

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

- choice_description column of the given dataset has 1246 missing values. 

### 2. Data Types:
#### Question: Verify the data types of each column. Do they align with their expected types, and should any adjustments be made?

In [5]:
df.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

- 'order_id' is basically the unique identification number used to uniquely identify each order. As it is a number it should be in the integer format

- 'quantity' refers to the number of items that have been requested or ordered. Hence it cannot be fractional, it should be in integer format

- 'item_name' is the unique name given to each dish to identify them based on there characteristics. Thus, it can be of string or of object type

- 'choice_description' represent the basic information of the dish for e.g. ingrediants, taste etc. Thus, it can be of string or of object type

- 'item_price' depends on the which currency it is mentioned. So, we have to format it based on the currency

### 3. Duplicated Entries:
#### Question: Identify and handle duplicated entries in the dataset. How might duplicates impact analysis, and what is the appropriate action?

In [6]:
df.duplicated().sum()

59

In [7]:
df = df.drop_duplicates(subset = ["order_id","quantity","item_name","choice_description"])

In [8]:
df.duplicated().sum()

0

- Duplicates can significantly impact the quality, accuracy, and reliability of your data, and lead to inaccurate results in your analysis or modeling.

- The appropriate action against duplicate date are
    - Ensure that your data is accurate and reliable
    - Prevent data duplicates in a dataset by creating a unique index
    - Data cleansing is important to this process, which involves removing duplicate information.

### 4. Quantity and Item Price:
#### Question: Examine the Quantity and Item Price columns. Are there any inconsistencies or anomalies that need correction?

### 9. Quantity and Price Relationships:
#### Question: Investigate the relationships between Quantity and Item Price. Are there cases where adjustments need to be made for accurate analysis?

In [9]:
df["quantity"].unique()

array([ 1,  2,  3,  4,  5, 15,  7,  8, 10], dtype=int64)

In [10]:
df["item_price"].unique()

array(['$2.39 ', '$3.39 ', '$16.98 ', '$10.98 ', '$1.69 ', '$11.75 ',
       '$9.25 ', '$4.45 ', '$8.75 ', '$11.25 ', '$8.49 ', '$2.18 ',
       '$8.99 ', '$1.09 ', '$2.95 ', '$2.15 ', '$3.99 ', '$22.50 ',
       '$11.48 ', '$17.98 ', '$17.50 ', '$4.30 ', '$5.90 ', '$1.25 ',
       '$23.78 ', '$6.49 ', '$11.08 ', '$1.50 ', '$22.16 ', '$32.94 ',
       '$22.20 ', '$10.58 ', '$2.50 ', '$23.50 ', '$7.40 ', '$18.50 ',
       '$3.00 ', '$6.78 ', '$11.89 ', '$9.39 ', '$4.00 ', '$3.75 ',
       '$8.69 ', '$2.29 ', '$8.90 ', '$3.27 ', '$3.89 ', '$8.19 ',
       '$35.00 ', '$27.75 ', '$11.80 ', '$6.00 ', '$26.25 ', '$21.96 ',
       '$4.36 ', '$7.50 ', '$4.78 ', '$13.35 ', '$6.45 ', '$5.07 ',
       '$22.96 ', '$7.17 ', '$7.98 ', '$4.50 ', '$26.07 ', '$12.98 ',
       '$35.25 ', '$44.25 ', '$10.50 ', '$33.75 ', '$16.38 ', '$13.52 ',
       '$5.00 ', '$15.00 ', '$8.50 ', '$17.80 ', '$1.99 ', '$11.49 '],
      dtype=object)

In [11]:
# formatting the salary label in our dataset
df["item_price"] = df["item_price"].str.replace("$","")

In [12]:
df["item_price"] = df["item_price"].astype(float).round(decimals = 2)

In [13]:
for i in df.index:
  if df.loc[i,"quantity"] >1:
    for j in df.index:
      if i == j:
        continue
      elif df.loc[i,"item_name"] == df.loc[j,"item_name"] and df.loc[i,"choice_description"] == df.loc[j,"choice_description"]:
        if df.loc[i,"item_price"] == df.loc[i,"quantity"]*df.loc[j,"item_price"]:
          continue
        else:
          df.loc[i,"item_price"] = df.loc[i,"quantity"]*df.loc[j,"item_price"]
          break

In [14]:
df.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
6                                                NaN        1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...        9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25

- There are no inconsistencies or anomalies in the 'quantity' label that may need any correction

- As Chipotle is american chain, it's default currency is $. Thus we can remove it while formatting the column

### 5. Choice Description:
#### Question: Analyze the Choice Description column. How should choices be handled, especially when there are multiple descriptions for a single item?

In [15]:
# np. set_printoptions(threshold=np. inf, suppress=True, linewidth=np. inf)
choice_discription_list = []
choice_discription_list = df["choice_description"].unique()
print(choice_discription_list)

[nan '[Clementine]' '[Apple]' ...
 '[Roasted Chili Corn Salsa, [Pinto Beans, Sour Cream, Cheese, Lettuce, Guacamole]]'
 '[Tomatillo Green Chili Salsa, [Rice, Black Beans]]'
 '[Tomatillo Green Chili Salsa, [Rice, Fajita Vegetables, Black Beans, Guacamole]]']


In [16]:
# Handling null values present in the choice_description label
df["choice_description"] = df["choice_description"].fillna("Default")

In [17]:
# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_columns', None)
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                            Default        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                            Default        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

- In choice_description column, assuming null values means 'we don't know the discription of that dish', Thus filling respective cell with the value as 'Unknown'

### 6. Handling Special Characters:
#### Question: Check for special characters in text-based columns (e.g., Item Name, Choice Description). How can these be addressed for consistency?

- first step was to remove special characters from item_name column

- concluding with removing '-' from choice_description column

In [18]:
df.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                            Default        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                            Default        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
6                                            Default        1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...        9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25

### 7. Order Id Integrity:
#### Question: Cross-reference the Order ID column for integrity. Are there any irregularities or patterns that need validation?

In [19]:
df.reset_index(inplace = True)

In [20]:
rows_to_drop = []

for x in range(len(df) - 1):
    if df.loc[x, "order_id"] <= df.loc[x + 1, "order_id"]:
        continue
    else:
        rows_to_drop.append(x)


df = df.drop(index=rows_to_drop).reset_index(drop=True)

In [21]:
df.head(10)
df.columns

Index(['index', 'order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

### 8. Item Name Standardization:
#### Question: Standardize the Item Name column. Are there variations that can be unified for better analysis?

In [22]:
item_name_list = []
item_name_list = df["item_name"].unique()
print(item_name_list)

['Chips and Fresh Tomato Salsa' 'Izze' 'Nantucket Nectar'
 'Chips and Tomatillo-Green Chili Salsa' 'Chicken Bowl' 'Side of Chips'
 'Steak Burrito' 'Steak Soft Tacos' 'Chips and Guacamole'
 'Chicken Crispy Tacos' 'Chicken Soft Tacos' 'Chicken Burrito'
 'Canned Soda' 'Barbacoa Burrito' 'Carnitas Burrito' 'Carnitas Bowl'
 'Bottled Water' 'Chips and Tomatillo Green Chili Salsa' 'Barbacoa Bowl'
 'Chips' 'Chicken Salad Bowl' 'Steak Bowl' 'Barbacoa Soft Tacos'
 'Veggie Burrito' 'Veggie Bowl' 'Steak Crispy Tacos'
 'Chips and Tomatillo Red Chili Salsa' 'Barbacoa Crispy Tacos'
 'Veggie Salad Bowl' 'Chips and Roasted Chili-Corn Salsa'
 'Chips and Roasted Chili Corn Salsa' 'Carnitas Soft Tacos'
 'Chicken Salad' 'Canned Soft Drink' 'Steak Salad Bowl'
 '6 Pack Soft Drink' 'Chips and Tomatillo-Red Chili Salsa' 'Bowl'
 'Burrito' 'Crispy Tacos' 'Carnitas Crispy Tacos' 'Steak Salad'
 'Chips and Mild Fresh Tomato Salsa' 'Veggie Soft Tacos'
 'Carnitas Salad Bowl' 'Barbacoa Salad Bowl' 'Salad' 'Veggie Cris

In [23]:
df["item_name"] = df["item_name"].str.replace("-"," ")

In [24]:
df.head(10)

index  order_id  quantity                              item_name  \
0      0         1         1           Chips and Fresh Tomato Salsa   
1      1         1         1                                   Izze   
2      2         1         1                       Nantucket Nectar   
3      3         1         1  Chips and Tomatillo Green Chili Salsa   
4      4         2         2                           Chicken Bowl   
5      5         3         1                           Chicken Bowl   
6      6         3         1                          Side of Chips   
7      7         4         1                          Steak Burrito   
8      8         4         1                       Steak Soft Tacos   
9      9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                            Default        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                            Default        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
6                                            Default        1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...        9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25

### 12. Handling Categorical Data:
#### Question: For categorical columns (e.g., Item Name), consider encoding or transforming them into a format suitable for analysis.

- There are no categorical data in the dataset

In [25]:
# Once the data is cleaned, exporting the cleaned data so that we can check its data integrity
df.to_csv('T2 - cleaned_data.csv')

### 10. Data Integrity Check:
#### Question: Perform a data integrity check by ensuring that quantities and prices align with the corresponding items and descriptions.

In [26]:
# Loading the cleaned data into pandas dataframe 
df_uncleaned = pd.read_csv("T2 - uncleaned_data.tsv" , sep = "\t")
df_cleaned = pd.read_csv("T2 - cleaned_data.csv")

In [27]:
#df_uncleaned.head()
print(df_cleaned.head())
print(df_uncleaned.head())

   Unnamed: 0  index  order_id  quantity  \
0           0      0         1         1   
1           1      1         1         1   
2           2      2         1         1   
3           3      3         1         1   
4           4      4         2         2   

                               item_name  \
0           Chips and Fresh Tomato Salsa   
1                                   Izze   
2                       Nantucket Nectar   
3  Chips and Tomatillo Green Chili Salsa   
4                           Chicken Bowl   

                                  choice_description  item_price  
0                                            Default        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                            Default        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
   order_id  quantity                              item_name  \
0   

In [28]:
print(df_uncleaned.columns)
print(df_cleaned.columns)

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')
Index(['Unnamed: 0', 'index', 'order_id', 'quantity', 'item_name',
       'choice_description', 'item_price'],
      dtype='object')


In [33]:
df_cleaned = df_cleaned.drop ('index', axis=1)
df_cleaned = df_cleaned.drop ('Unnamed: 0', axis=1)

KeyError: "['index'] not found in axis"

In [34]:
#Compare the shapes, columns, and data types of the DataFrames
print("Shape of uncleaned data:", df_uncleaned.shape)
print("Shape of cleaned data:", df_cleaned.shape)
print()

print("Columns of uncleaned data:", df_uncleaned.columns)
print("Columns of cleaned data:", df_cleaned.columns)
print()

print("Data types of uncleaned data:")
print(df_uncleaned.dtypes)
print()

print("Data types of cleaned data:")
print(df_cleaned.dtypes)
print()

Shape of uncleaned data: (4622, 5)
Shape of cleaned data: (4563, 6)

Columns of uncleaned data: Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')
Columns of cleaned data: Index(['Unnamed: 0', 'order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

Data types of uncleaned data:
order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

Data types of cleaned data:
Unnamed: 0              int64
order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object



In [35]:
# Summarize the statistics of the DataFrames
print("Summary of uncleaned data:")
print(df_uncleaned.describe())
print()

print("Summary of cleaned data:")
print(df_cleaned.describe())
print()

Summary of uncleaned data:
          order_id     quantity
count  4622.000000  4622.000000
mean    927.254868     1.075725
std     528.890796     0.410186
min       1.000000     1.000000
25%     477.250000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

Summary of cleaned data:
        Unnamed: 0     order_id     quantity   item_price
count  4563.000000  4563.000000  4563.000000  4563.000000
mean   2281.000000   927.045803     1.076704     8.041034
std    1317.368969   528.936100     0.412739     7.720611
min       0.000000     1.000000     1.000000     1.090000
25%    1140.500000   476.500000     1.000000     3.990000
50%    2281.000000   926.000000     1.000000     8.750000
75%    3421.500000  1393.000000     1.000000    10.780000
max    4562.000000  1834.000000    15.000000   211.500000



In [37]:
# Check for any missing or duplicated values in the DataFrames
print("Missing values in uncleaned data:")
print(df_uncleaned.isnull().sum())
print()

print("Missing values in cleaned data:")
print(df_cleaned.isnull().sum())
print()

print("Duplicated values in uncleaned data:")
print(df_uncleaned.duplicated().sum())
print()

print("Duplicated values in cleaned data:")
print(df_cleaned.duplicated().sum())
print()

Missing values in uncleaned data:
order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

Missing values in cleaned data:
Unnamed: 0            0
order_id              0
quantity              0
item_name             0
choice_description    0
item_price            0
dtype: int64

Duplicated values in uncleaned data:
59

Duplicated values in cleaned data:
0



### 11. Converting to CSV:
#### Optional Question: If needed, convert the cleaned dataset to a CSV file using the to_csv function with sep=','.

In [38]:
df_csv = df.to_csv("T1 - cleaned_data.tsv",sep = ',')